src/main/java/com/sk89q/craftbook/mechanics/pipe/Pipes.java

```diff
@@ -39,16 +39,10 @@
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.util.Vector;
 
-import java.util.ArrayDeque;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.Deque;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
 
 public class Pipes extends AbstractCraftBookMechanic {
+    private Map<String, Long> lastFailedAttempt = new HashMap<>();
```

Comment:
Is this guaranteed to be cleared of all added entries? This could easily cause a pretty major runaway memory problem